\documentclass{article}

\usepackage{graphicx}
\usepackage{afterpage}
\usepackage[margin=0.1in]{geometry}
\renewcommand{\figurename}{Supplementary Figure}
\begin{document}
\renewcommand{\floatpagefraction}{0.1}

\clearpage
\afterpage{
    \begin{figure}[p]
        \centering
        \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure1.xy_strata_4_5}
        \caption{(A) Visualization with Saffire (https://mrvollger.github.io/SafFire/) of the alignment between T2T-CHM13 X and Y reveals that strata 5 and 4 feature low identity (~90\%), numerous inversions, and some rearrangements; (B) X chromosome ideogram according to (Ross et al. 2005). On the bottom, its evolutionary domains: the X-added region (XAR), the X-conserved region (XCR; dotted region in proximal Xp does not appear to be part of the XCR), the pseudoautosomal region PAR1, and evolutionary strata S5–S1. (C) The reduced all-to-all mapping graph of HPRCy1 versus itself, with contigs represented as nodes and mappings as edges, rendered in Gephi (Bastian, Heymann, and Jacomy 2009). In red contigs covering the evolutionary strata 5 and 4 on chromosome X; (D) Coloring the reduced homology mapping graph in C with community assignments. Panels C and D use the same layout as Figure 1 but focus only on the X and Y region of the visualization.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure2}
    \caption{An overview of our approach to build a PVG for HPRCy1 contigs that can be anchored to a specific acrocentric q-arm. (A) As input, we take the entire HPRCy1 and map it to T2T-CHM13. (B) This yields mappings to acrocentric chromosomes, which we filter to select contigs that map across the centromeres (red cytobands) between non-centromeric regions (over-labeled green). We include two HG002 assemblies based on standard HiFi (from HPRCy1) and on both HiFi and ONT data (from Verkko). (C) We then apply PGGB to build a PVG from the HPRCy1-acro collection. PGGB first obtains an all-to-all alignment of the input (C.a.), which is converted to a variation graph with SEQWISH (Garrison and Guarracino 2022) (C.b.), then normalized with sorting and multiple sequence alignment steps in SMOOTHXG (C.c-f). (D) The resulting PVG expresses genomes as paths, or walks, through a common sequence graph. This model thus contains all input sequences and their relative alignments to all others—in the example we see a CTGG/AAGTA block substitution between genomes 1 and 2.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure3.SegDupWithSST1array}
    \caption{Regions of a PVG built from centromere-spanning HPRCy1 contigs plus T2T-CHM13 and GRCh38 references. We apply GFAESTUS to visualize the 2D layout generated by ODGI in the PGGB pipeline. We focus on the segmentally duplicated core centered in the SST1 array (labeled). }
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure4.chrACRO.pq_contigs.1kbps.hg002prox.hg002hifi.fa.gz.31257a6.04f1c29.eb4a615.smooth.final.og.lay.draw}
    \caption{A PVG built from centromere-spanning HPRCy1 contigs, without embedding the T2T-CHM13 and GRCh38 references. We apply ODGI to visualize the 2D layout generated in the PGGB pipeline. This renders sequences and chains of small variants as linear structures, while repeats caused by segmental duplications, inversions, and other structural variants tend to form loops and tangles. The acrocentric q-arms are almost completely separated, while the p-arms unite in a structure adjacent to the rDNA array.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure5}
    \caption{Regions of a PVG built from centromere-spanning HPRCy1 contigs without the T2T-CHM13 and GRCh38 references. We apply GFAESTUS to visualize the 2D layout generated by ODGI in the PGGB pipeline. We focus on the segmentally duplicated core centered in the SST1 array (labeled).}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure6.Untangling}
    \caption{Scheme of the graph untangling. We applied ODGI UNTANGLE to obtain a mapping from segments of all PVG paths onto T2T-CHM13. The segmentation cuts the graph into regular-sized regions whose boundaries occur at structural variant breakpoints. For each query subpath through a graph segment, we use a Jaccard metric over the sequence space of the subpaths to find the best-matching reference segment.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure7.e50000.m1000.annot.chr13}
    \caption{Untangling of HPRCy1-acro’s sequences belonging to chromosome 13 versus T2T-CHM13. We display all mappings above 90\% estimated pairwise identity, without removing those covering regions classified as unreliable. Above each contig, we display the unreliable regions in black.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure8.e50000.m1000.annot.chr14}
    \caption{Untangling of HPRCy1-acro’s sequences belonging to chromosome 14 versus T2T-CHM13. We display all mappings above 90\% estimated pairwise identity, without removing those covering regions classified as unreliable. Above each contig, we display the unreliable regions in black.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure9.e50000.m1000.annot.chr15}
    \caption{Untangling of HPRCy1-acro’s sequences belonging to chromosome 15 versus T2T-CHM13. We display all mappings above 90\% estimated pairwise identity, without removing those covering regions classified as unreliable. Above each contig, we display the unreliable regions in black.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure10.e50000.m1000.annot.chr21}
    \caption{Untangling of HPRCy1-acro’s sequences belonging to chromosome 21 versus T2T-CHM13. We display all mappings above 90\% estimated pairwise identity, without removing those covering regions classified as unreliable. Above each contig, we display the unreliable regions in black.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure11.e50000.m1000.annot.chr22}
    \caption{Untangling of HPRCy1-acro’s sequences belonging to chromosome 22 versus T2T-CHM13. We display all mappings above 90\% estimated pairwise identity, without removing those covering regions classified as unreliable. Above each contig, we display the unreliable regions in black.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure12.CenSatAnnotationTrackLegend}
    \caption{Centromeric Satellite Annotation (CenSat Annotation) track legend.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure13.chr13}
    \caption{(A) We focus on the first 25 Mbp of chromosome 13 shown here as a red box over T2T-CHM13 cytobands. Pseudo-homologous regions (PHRs), where diverse sets of acrocentric chromosomes recombine, are highlighted relative to T2T-CHM13 genome annotations for repeats, GC percentage, and genes. Above, we indicate regions of interest described in the main text: rDNA, SST1 array, centromere, and q-arm. Below, we show T2T-CHM13-relative homology mosaics for each chromosome 13 matched contig from HPRCy1-acro, with the most-similar reference chromosome at each region shown using the given colors (Target). (B) Aggregated untangle results in the SAACs. For each acrocentric chromosome, we show the count of its HPRCy1 q-arm-anchored contigs mapping itself and all other acrocentrics (Contigs), (C) as well as the regional (50kbp) untangle entropy metric (Regional homology entropy) computed over the contigs’ T2T-CHM13-relative untanglings. (D) By considering the multiple untangling of each HPRCy1-acro contig, we develop a point-wise metric that captures diversity in T2T-CHM13-relative homology patterns (Positional homology entropy), leading to our definition of the PHRs. (E) The patterns of homology mosaicism suggest ongoing recombination exchange in the SAACs. A scan over T2T-CHM13 reveals that the rDNA units are enriched for PRDM9 binding motifs, and thus may host frequent double stranded breaks during meiosis. In (B-D) a gray background indicates regions with missing data due to the lack of non-T2T-CHM13 contigs. We provide the Centromeric Satellite Annotation (CenSat Annotation) track legend in Supplementary Figure 12.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure14.chr14}
    \caption{(A) We focus on the first 25 Mbp of chromosome 14 shown here as a red box over T2T-CHM13 cytobands. Pseudo-homologous regions (PHRs), where diverse sets of acrocentric chromosomes recombine, are highlighted relative to T2T-CHM13 genome annotations for repeats, GC percentage, and genes. Above, we indicate regions of interest described in the main text: rDNA, SST1 array, centromere, and q-arm. Below, we show T2T-CHM13-relative homology mosaics for each chromosome 13 matched contig from HPRCy1-acro, with the most-similar reference chromosome at each region shown using the given colors (Target). (B) Aggregated untangle results in the SAACs. For each acrocentric chromosome, we show the count of its HPRCy1 q-arm-anchored contigs mapping itself and all other acrocentrics (Contigs), (C) as well as the regional (50kbp) untangle entropy metric (Regional homology entropy) computed over the contigs’ T2T-CHM13-relative untanglings. (D) By considering the multiple untangling of each HPRCy1-acro contig, we develop a point-wise metric that captures diversity in T2T-CHM13-relative homology patterns (Positional homology entropy), leading to our definition of the PHRs. (E) The patterns of homology mosaicism suggest ongoing recombination exchange in the SAACs. A scan over T2T-CHM13 reveals that the rDNA units are enriched for PRDM9 binding motifs, and thus may host frequent double stranded breaks during meiosis. In (B-D) a gray background indicates regions with missing data due to the lack of non-T2T-CHM13 contigs. We provide the Centromeric Satellite Annotation (CenSat Annotation) track legend in Supplementary Figure 12.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure15.chr15}
    \caption{(A) We focus on the first 25 Mbp of chromosome 15 shown here as a red box over T2T-CHM13 cytobands. Pseudo-homologous regions (PHRs), where diverse sets of acrocentric chromosomes recombine, are highlighted relative to T2T-CHM13 genome annotations for repeats, GC percentage, and genes. Above, we indicate regions of interest described in the main text: rDNA, SST1 array, centromere, and q-arm. Below, we show T2T-CHM13-relative homology mosaics for each chromosome 13 matched contig from HPRCy1-acro, with the most-similar reference chromosome at each region shown using the given colors (Target). (B) Aggregated untangle results in the SAACs. For each acrocentric chromosome, we show the count of its HPRCy1 q-arm-anchored contigs mapping itself and all other acrocentrics (Contigs), (C) as well as the regional (50kbp) untangle entropy metric (Regional homology entropy) computed over the contigs’ T2T-CHM13-relative untanglings. (D) By considering the multiple untangling of each HPRCy1-acro contig, we develop a point-wise metric that captures diversity in T2T-CHM13-relative homology patterns (Positional homology entropy), leading to our definition of the PHRs. (E) The patterns of homology mosaicism suggest ongoing recombination exchange in the SAACs. A scan over T2T-CHM13 reveals that the rDNA units are enriched for PRDM9 binding motifs, and thus may host frequent double stranded breaks during meiosis. In (B-D) a gray background indicates regions with missing data due to the lack of non-T2T-CHM13 contigs. We provide the Centromeric Satellite Annotation (CenSat Annotation) track legend in Supplementary Figure 12.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure16.chr21}
    \caption{(A) We focus on the first 25 Mbp of chromosome 21 shown here as a red box over T2T-CHM13 cytobands. Pseudo-homologous regions (PHRs), where diverse sets of acrocentric chromosomes recombine, are highlighted relative to T2T-CHM13 genome annotations for repeats, GC percentage, and genes. Above, we indicate regions of interest described in the main text: rDNA, SST1 array, centromere, and q-arm. Below, we show T2T-CHM13-relative homology mosaics for each chromosome 13 matched contig from HPRCy1-acro, with the most-similar reference chromosome at each region shown using the given colors (Target). (B) Aggregated untangle results in the SAACs. For each acrocentric chromosome, we show the count of its HPRCy1 q-arm-anchored contigs mapping itself and all other acrocentrics (Contigs), (C) as well as the regional (50kbp) untangle entropy metric (Regional homology entropy) computed over the contigs’ T2T-CHM13-relative untanglings. (D) By considering the multiple untangling of each HPRCy1-acro contig, we develop a point-wise metric that captures diversity in T2T-CHM13-relative homology patterns (Positional homology entropy), leading to our definition of the PHRs. (E) The patterns of homology mosaicism suggest ongoing recombination exchange in the SAACs. A scan over T2T-CHM13 reveals that the rDNA units are enriched for PRDM9 binding motifs, and thus may host frequent double stranded breaks during meiosis. In (B-D) a gray background indicates regions with missing data due to the lack of non-T2T-CHM13 contigs. We provide the Centromeric Satellite Annotation (CenSat Annotation) track legend in Supplementary Figure 12.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure17.chr22}
    \caption{(A) We focus on the first 25 Mbp of chromosome 22 shown here as a red box over T2T-CHM13 cytobands. Pseudo-homologous regions (PHRs), where diverse sets of acrocentric chromosomes recombine, are highlighted relative to T2T-CHM13 genome annotations for repeats, GC percentage, and genes. Above, we indicate regions of interest described in the main text: rDNA, SST1 array, centromere, and q-arm. Below, we show T2T-CHM13-relative homology mosaics for each chromosome 13 matched contig from HPRCy1-acro, with the most-similar reference chromosome at each region shown using the given colors (Target). (B) Aggregated untangle results in the SAACs. For each acrocentric chromosome, we show the count of its HPRCy1 q-arm-anchored contigs mapping itself and all other acrocentrics (Contigs), (C) as well as the regional (50kbp) untangle entropy metric (Regional homology entropy) computed over the contigs’ T2T-CHM13-relative untanglings. (D) By considering the multiple untangling of each HPRCy1-acro contig, we develop a point-wise metric that captures diversity in T2T-CHM13-relative homology patterns (Positional homology entropy), leading to our definition of the PHRs. (E) The patterns of homology mosaicism suggest ongoing recombination exchange in the SAACs. A scan over T2T-CHM13 reveals that the rDNA units are enriched for PRDM9 binding motifs, and thus may host frequent double stranded breaks during meiosis. In (B-D) a gray background indicates regions with missing data due to the lack of non-T2T-CHM13 contigs. We provide the Centromeric Satellite Annotation (CenSat Annotation) track legend in Supplementary Figure 12.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure18}
    \caption{For each base position of T2T-CHM13 chromosome 13, we compared the untangling best-hit of the HG002-HPRCy1 contigs with the best-hit supported by HG002-Verkko’s contigs. We considered only best-hits with an estimated identity of at least 90\% that cover regions labeled as reliable. We reported the number of different targets for each reference position. On the bottom, the untangling of HG002’s contigs from HG002-Verkko assembly as those seen in HG002-HPRCy1 assemblies, for chromosome 13 versus T2T-CHM13. Transparency shows the estimated identity of the mappings. We display all mappings above 90\% estimated pairwise identity. Checkerboard patterns observed in several regions of the SAACs correspond to contexts that may permit ongoing recombination.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure19}
    \caption{For each base position of T2T-CHM13 chromosome 14, we compared the untangling best-hit of the HG002-HPRCy1 contigs with the best-hit supported by HG002-Verkko’s contigs. We considered only best-hits with an estimated identity of at least 90\% that cover regions labeled as reliable. We reported the number of different targets for each reference position. On the bottom, the untangling of HG002’s contigs from HG002-Verkko assembly as those seen in HG002-HPRCy1 assemblies, for chromosome 14 versus T2T-CHM13. Transparency shows the estimated identity of the mappings. We display all mappings above 90\% estimated pairwise identity. Checkerboard patterns observed in several regions of the SAACs correspond to contexts that may permit ongoing recombination.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure20}
    \caption{For each base position of T2T-CHM13 chromosome 15, we compared the untangling best-hit of the HG002-HPRCy1 contigs with the best-hit supported by HG002-Verkko’s contigs. We considered only best-hits with an estimated identity of at least 90\% that cover regions labeled as reliable. We reported the number of different targets for each reference position. On the bottom, the untangling of HG002’s contigs from HG002-Verkko assembly as those seen in HG002-HPRCy1 assemblies, for chromosome 15 versus T2T-CHM13. Transparency shows the estimated identity of the mappings. We display all mappings above 90\% estimated pairwise identity. Checkerboard patterns observed in several regions of the SAACs correspond to contexts that may permit ongoing recombination.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure21}
    \caption{For each base position of T2T-CHM13 chromosome 21, we compared the untangling best-hit of the HG002-HPRCy1 contigs with the best-hit supported by HG002-Verkko’s contigs. We considered only best-hits with an estimated identity of at least 90\% that cover regions labeled as reliable. We reported the number of different targets for each reference position. On the bottom, the untangling of HG002’s contigs from HG002-Verkko assembly as those seen in HG002-HPRCy1 assemblies, for chromosome 21 versus T2T-CHM13. Transparency shows the estimated identity of the mappings. We display all mappings above 90\% estimated pairwise identity. Checkerboard patterns observed in several regions of the SAACs correspond to contexts that may permit ongoing recombination.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure22}
    \caption{For each base position of T2T-CHM13 chromosome 22, we compared the untangling best-hit of the HG002-HPRCy1 contigs with the best-hit supported by HG002-Verkko’s contigs. We considered only best-hits with an estimated identity of at least 90\% that cover regions labeled as reliable. We reported the number of different targets for each reference position. On the bottom, the untangling of HG002’s contigs from HG002-Verkko assembly as those seen in HG002-HPRCy1 assemblies, for chromosome 22 versus T2T-CHM13. Transparency shows the estimated identity of the mappings. We display all mappings above 90\% estimated pairwise identity. Checkerboard patterns observed in several regions of the SAACs correspond to contexts that may permit ongoing recombination.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[scale=0.21,keepaspectratio]{SupplementaryFigure23.chr13.SST1.1Mbps.n1.nref1}
    \caption{Multiple untangling of T2T-CHM13, GRCh38, HG002-Verkko haplotypes, and HPRCy1-acro contigs versus T2T-CHM13. Chromosome 13 results are represented, in the chr13:11,301,367-13,440,010 region (censat\_13\_27 coordinates +- 1Mbp). Transparency shows the different orientations of the mappings. We display all mappings above 90\% estimated pairwise identity. To analyze simultaneous hits to all acrocentrics, each grouping shows the first 3 best alternative mappings. The figure shows that chromosome 13’s contigs map in forward orientation on T2T-CHM13 chromosome 13 and 21 (orange and cyan rectangles), while their mappings are inverted on chromosomes 14 (transparent gold rectangles).}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[scale=0.21,keepaspectratio]{SupplementaryFigure24.chr14.SST1.1Mbps.n1.nref1}
    \caption{Multiple untangling of T2T-CHM13, GRCh38, HG002-Verkko haplotypes, and HPRCy1-acro contigs versus T2T-CHM13. Chromosome 14 results are represented, in the chr14:5,960,008-7,988,409 region (censat\_14\_39 coordinates +- 1Mbp). Transparency shows the different orientations of the mappings. We display all mappings above 90\% estimated pairwise identity. To analyze simultaneous hits to all acrocentrics, each grouping shows the first 3 best alternative mappings. The figure shows that chromosome 14’s contigs map in forward orientation on T2T-CHM13 chromosome 14 (gold rectangles), while their mappings are inverted on chromosomes 13 and 21 (transparent orange and cyan rectangles), with the sole exception of HG01071\#2\#JAHBCE010000082.1, where the trend is reversed.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[scale=0.21,keepaspectratio]{SupplementaryFigure25.chr21.SST1.1Mbps.n1.nref1}
    \caption{Multiple untangling of T2T-CHM13, GRCh38, HG002-Verkko haplotypes, and HPRCy1-acro contigs versus T2T-CHM13. Chromosome 21 results are represented, in the chr21:8,375,567-10,453,313 region (censat\_21\_45 coordinates +- 1Mbp). Transparency shows the different orientations of the mappings. We display all mappings above 90\% estimated pairwise identity. To analyze simultaneous hits to all acrocentrics, each grouping shows the first 3 best alternative mappings. The figure shows that chromosome 21’s contigs map in forward orientation on T2T-CHM13 chromosome 21 and 13 (cyan and orange rectangles), while their mappings are inverted on chromosomes 14 (transparent gold rectangles).}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure26}
    \caption{For HPRCy1 contigs from chromosome X and Y, we show the tracks for the pseudoautosomal regions (PARs) and the X-transposed region (XTRs) with respect to T2T-CHM13 (on the top) as well as the untangle entropy metric (Regional homology metric, on the bottom) computed over the contigs’ T2T-CHM13-relative untanglings.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[scale=0.16,keepaspectratio]{SupplementaryFigure27}
    \caption{Multiple untangling of HPRCy1-acro’s sequences belonging to chromosome 13 versus T2T-CHM13. Transparency shows the estimated identity of the mappings. We display all mappings above 90\% estimated pairwise identity that cover regions labeled as reliable. To allow the display of simultaneous hits to all acrocentrics, each grouping shows the first 5 best alternative mappings. Checkerboard patterns observed in several regions of the SAACs correspond to contexts that may permit ongoing recombination.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[scale=0.16,keepaspectratio]{SupplementaryFigure28}
    \caption{Multiple untangling of HPRCy1-acro’s sequences belonging to chromosome 14 versus T2T-CHM13. Transparency shows the estimated identity of the mappings. We display all mappings above 90\% estimated pairwise identity that cover regions labeled as reliable. To allow the display of simultaneous hits to all acrocentrics, each grouping shows the first 5 best alternative mappings. Checkerboard patterns observed in several regions of the SAACs correspond to contexts that may permit ongoing recombination.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[scale=0.16,keepaspectratio]{SupplementaryFigure29}
    \caption{Multiple untangling of HPRCy1-acro’s sequences belonging to chromosome 15 versus T2T-CHM13. Transparency shows the estimated identity of the mappings. We display all mappings above 90\% estimated pairwise identity that cover regions labeled as reliable. To allow the display of simultaneous hits to all acrocentrics, each grouping shows the first 5 best alternative mappings. Checkerboard patterns observed in several regions of the SAACs correspond to contexts that may permit ongoing recombination.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[scale=0.16,keepaspectratio]{SupplementaryFigure30}
    \caption{Multiple untangling of HPRCy1-acro’s sequences belonging to chromosome 21 versus T2T-CHM13. Transparency shows the estimated identity of the mappings. We display all mappings above 90\% estimated pairwise identity that cover regions labeled as reliable. To allow the display of simultaneous hits to all acrocentrics, each grouping shows the first 5 best alternative mappings. Checkerboard patterns observed in several regions of the SAACs correspond to contexts that may permit ongoing recombination.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[scale=0.16,keepaspectratio]{SupplementaryFigure31}
    \caption{Multiple untangling of HPRCy1-acro’s sequences belonging to chromosome 22 versus T2T-CHM13. Transparency shows the estimated identity of the mappings. We display all mappings above 90\% estimated pairwise identity that cover regions labeled as reliable. To allow the display of simultaneous hits to all acrocentrics, each grouping shows the first 5 best alternative mappings. Checkerboard patterns observed in several regions of the SAACs correspond to contexts that may permit ongoing recombination.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure32}
    \caption{Multi-hit untangling diversity entropy for HPRCy1-acro’s sequences belonging to chromosome 13, 14, 15, 21, 22 versus T2T-CHM13. We considered all mappings above 90\% estimated pairwise identity that cover regions labeled as reliable.
}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure33}
    \caption{For each T2T-CHM13 acrocentric chromosome, we show the tracks for the pseudo-homologous regions (PHRs), the SST1 array, the rDNA arrays, and for the regions where the bacterial artificial chromosome (BAC) clones from (Jarmuz-Szymczak et al. 2014) map on those chromosomes. Most of the mappings cover the PHRs. The mappings with higher estimated identity (> 99\%) are on chr21 and chr14. We colored BAC clones’ mappings according to (Jarmuz-Szymczak et al. 2014).}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[scale=2,keepaspectratio]{SupplementaryFigure34.SST1_consensus}
    \caption{Comparison of chr13 and chr17 SST1 consensus sequences by alignment dotplot indicate the unique deletion shared exclusively by the acrocentric chromosomes (dotted red rectangle).}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure35}
    \caption{For HPRCy1 contigs from chromosome X and Y, we show the tracks for the pseudoautosomal regions (PARs) and the X-transposed region (XTRs) with respect to T2T-CHM13 as well as the multi untangle entropy metric (Positional homology entropy, on the bottom) computed over the contigs’ T2T-CHM13-relative untanglings.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure36}
    \caption{Biallelic SNP density in the full HPRCy1 draft pangenome. Colors indicate the count of SNPs in each bin at given variant nesting levels (Paten et al. 2018), with gray showing the base level and a red to violet spectrum for increasing level.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure37.chm13v2.PRDM9.chrACRO.w20000}
    \caption{For each T2T-CHM13 acrocentric chromosome, we show the number of human PRDM9 binding motif hits present in windows 20 kbps long.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure38.genotype_stats}
    \caption{Left) Histograms showing missing genotypes per site for variants called with respect to the T2T-CHM13 acrocentrics. Middle) Histograms showing the allele frequency spectrum for variants called with respect to the T2T-CHM13 acrocentrics. MAF = Minor Allele Frequency. Right-Top) Segment size distributions of the pseudo-homologous regions with respect to the T2T-CHM13 acrocentrics. Right-Bottom) Distributions of the distance between consecutive markers with respect to the T2T-CHM13 acrocentrics.}
    \end{figure}
}
\clearpage
\afterpage{
    \begin{figure}[p]
    \centering
    \includegraphics[width=\textwidth,height=\textheight,keepaspectratio]{SupplementaryFigure39.HPRCy1v2genbank.self.s10k.l50k.p95.n93.h0001.l1000000.paf.community.leiden.composition}
    \caption{Results of community assignment on the mapping graph when using chains of 10 kbp seeds of 95\% average nucleotide identity. On the x-axis the chromosome to which contigs belong based on competitive mapping to T2T-CHM13 and GRCh38, while the y-axis indicates the community, which is named by the chromosome partition that contributes the largest number of contigs to it.}
    \end{figure}
}
\clearpage

\end{document}
